### PR TITLE
Converts existing native histogram unit tests to the PromQL testing framework

### LINKED ISSUE #13563: Convert existing native histogram unit tests to the PromQL testing framework
### Proposal

The PromQL testing framework has been upgraded to support native histograms, but we still have a lot of histogram tests written as code level unit tests that we can now convert into much simpler tests within the PromQL testing framework.

In other words: More tests need to be added to https://github.com/prometheus/prometheus/blob/main/promql/testdata/histograms.test that do exactly the same as the following unit tests (which we can then delete):

- [x] [TestNativeHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3155)
- [x] [TestNativeFloatHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3243C6-L3243C34)
- [x] [TestNativeHistogram_HistogramCountAndSum](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3284C6-L3284C46)
- [x] [TestNativeHistogram_HistogramStdDevVar](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3363C6-L3363C44)
- [x] [TestNativeHistogram_HistogramQuantile](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3522C6-L3522C43)
- [x] [TestNativeHistogram_HistogramFraction](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3758C6-L3758C43)
- [ ] [TestNativeHistogram_Sum_Count_Add_AvgOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4193C6-L4193C51)
- [ ] [TestNativeHistogram_SubOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4388C6-L4388C37)
- [ ] [TestNativeHistogram_MulDivOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4623C6-L4623C40)

### PR DESCRIPTION
**Pull request by NeerajGartia21**
Fixes #13563 

This is currently a draft PR to check whether I'm in the right path. After review I'll add subsequent tests.

@beorn7 can you please review.

### CODE REVIEW

**Comment by beorn7** [comment]
Thanks, I'll have a look ASAP.

**Review by beorn7 — CHANGES_REQUESTED** [review]
This is broadly the right direction, but let's try to mirror the original test more closely, see comments.

File: promql/testdata/native_histograms.test
Comment (NeerajGartia21) [reply promql/testdata/native_histograms.test:275]: Got it. Thanks !!
I'll try to replicate the `GenerateTestHistogram` functionality for the test.

**Comment by beorn7** [comment]
@NeerajGartia21 are you still working on this? Do you need more help?

**Comment by NeerajGartia21** [comment]
Yes, I'm working on it but I've got an issue, don't know whether I'm wrong or it's a bug.  But I've written a test for the NativeHistogramRate which is as follows-

```
load 15s
    test_histogram {{schema:1 count:12 sum:18.4 z_bucket:2 z_bucket_w:0.001 buckets:[1 2 0 1 1] n_buckets:[1 2 0 1 1]}}+{{schema:1 count:9 sum:18.4 z_bucket:1 z_bucket_w:0.001 buckets:[1 1 0 1 1] n_buckets:[1 1 0 1 1]}}x100

eval instant at 5m rate(test_histogram[45s])
    {} {{schema:1 count:0.6 sum:1.2266666666666652 z_bucket:0.06666666666666667 z_bucket_w:0.001 buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667] n_buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667]}}
```
I'm getting error on this but as you can see here the expected and actual values are exactly same. I'm trying to debug it but I need some help.
```
error in eval rate(test_histogram[45s]) (line 277): expected 
{{
schema:1 
count:0.6 
sum:1.2266666666666652 
z_bucket:0.06666666666666667 
z_bucket_w:0.001 
buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667] 
n_buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667]
}} 
for {} but got 
{{
schema:1 
count:0.6 
sum:1.2266666666666652 
z_bucket:0.06666666666666667 
z_bucket_w:0.001 
buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667] 
n_buckets:[0.06666666666666667 0.06666666666666667 0 0.06666666666666667 0.06666666666666667]
}}
```
P.S- I've formatted the error for better visibility.

**Comment by beorn7** [comment]
That looks weird indeed.

Could you push your current version of the code (that runs into the error above)? Then I might take part in the mystery hunt (if I find some time…).

**Comment by NeerajGartia21** [comment]
Yeah, It's weird actually. I've pushed the code. I'm trying to find the bug for this/

**Comment by beorn7** [comment]
I found the problem, https://github.com/prometheus/prometheus/pull/13854. Please review. :)

**Comment by beorn7** [comment]
BTW: The histogram comparison relies on exact float matches rather using `almostEqual`. If that turns out to be a problem with any of the other tests, we should keep that in mind. (It works fine for this test.)

**Comment by NeerajGartia21** [comment]
> I found the problem, #13854. Please review. :)

Sure! Thanks. I'll review it.

> BTW: The histogram comparison relies on exact float matches rather using `almostEqual`. If that turns out to be a problem with any of the other tests, we should keep that in mind. (It works fine for this test.)

Yeah, we should keep that in mind.

**Comment by beorn7** [comment]
OK, so after a rebase, you should get this working.

**Comment by NeerajGartia21** [comment]
Yeah, I can start working it now. Thanks @beorn7 .

**Comment by beorn7** [comment]
I'll try to get to this ASAP, but currently, I'm sick.

**Comment by NeerajGartia21** [comment]
No problem. Take care @beorn7. 
Sorry, I was also busy with my GSoC proposal for the `created_timestamp` one and couldn't look into it. I'll add all the other tests by tomorrow.

**Comment by NeerajGartia21** [comment]
To support buckets with negative values like- `buckets:[2 1 -2 3]`, we need to change the grammar from 
```
bucket_set_list : bucket_set_list SPACE number
                {
                  $$ = append($1, $3)
                }
                | number
                {
                  $$ = []float64{$1}
                }
                | bucket_set_list error
                ;
```
to 
```
bucket_set_list : bucket_set_list SPACE signed_or_unsigned_number
                {
                  $$ = append($1, $3)
                }
                | signed_or_unsigned_number
                {
                  $$ = []float64{$1}
                }
                | bucket_set_list error
                ;
```
but when I'm trying to generate the grammar I'm getting `conflicts: 19 shift/reduce` error. @beorn7 Can I get some help with this?

**Comment by NeerajGartia21** [comment]
@beorn7 is there any way to write these values in promql testing framework without writting these many zeroes?
```
				PositiveSpans: []histogram.Span{
					{Offset: 0, Length: 1},
					{Offset: 255, Length: 1},
					{Offset: 149, Length: 1},
					{Offset: 105, Length: 1},
				},
				PositiveBuckets: []int64{1, 0, 0, 0},
```

**Comment by beorn7** [comment]
No, there isn't. Back then, we decided that the grammar of the testing framework doesn't have to be that powerful. 

This particular test might be one that we don't want to convert. Or we change it slightly to have fewer gaps.

If the test is specifically about testing large gaps, then maybe it should stay as it is.

**Comment by NeerajGartia21** [comment]
This test is about testing `histogram_stddev` and `histogram_stdvar` with high resolution histogram with values. So should I change it with smaller gaps or add these many zeroes?

**Comment by beorn7** [comment]
Try it with smaller gaps, if you can still prove the point (that it estimates the stddev/stdvar correctly). Maybe the original gaps were chosen because it make the calculation very simple.

**Comment by NeerajGartia21** [comment]
Okay, I'll try with smaller gaps.

**Comment by NeerajGartia21** [comment]
@beorn7 Is there any way to support `Inf` or `NaN` for expressions like `Sum: math.NaN()`  in current promql testing framework? Or Do I need to add `Inf` to `generate_parser.y` ?

**Comment by NeerajGartia21** [comment]
Yeah it should support these values and currently Idt it supports these values. I'll try to look into it and will create a PR subsequently. But I might need some help in compiling the `generate_parser.y` file as I encountered some error while running `make parser` command.

**Comment by NeerajGartia21** [comment]
Created a PR for this issue in #13916 .

**Comment by NeerajGartia21** [comment]
Just a quick update. I've added tests for half of the functions which are-
- [TestNativeHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3155)
- [TestNativeFloatHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3243C6-L3243C34)
- [TestNativeHistogram_HistogramCountAndSum](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3284C6-L3284C46)
- [TestNativeHistogram_HistogramStdDevVar](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3363C6-L3363C44)
- [TestNativeHistogram_HistogramQuantile](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3522C6-L3522C43)
- [TestNativeHistogram_HistogramFraction](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3758C6-L3758C43)

I'll add tests for other ASAP. 
Also I assume in promql testing framework Histograms are by default FloatHistogram, So I think `TestNativeHistogramRate` and `TestFloatHistogramRate` will be same. Please correct me if I'm wrong.

**Comment by beorn7** [comment]
Thank you very much. I'm on vacation this week, but this is high on my list once I'm back in action.

File: promql/testdata/native_histograms.test
Comment (beorn7) [inline promql/testdata/native_histograms.test:716]: Newline missing here.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thank you very much.

Could you also remove the conventional unit tests that the new framework tests are replacing? In that way, it's easier for me to double-check if they are testing the same thing. And we have to remove the unit tests eventually, anyway.

**Comment by beorn7** [comment]
Note that #13999 might get merged very soon, and it will create merge conflicts. I think they will be easy to solve.

**Comment by NeerajGartia21** [comment]
> Thank you very much.
> Could you also remove the conventional unit tests that the new framework tests are replacing? In that way, it's easier for me to double-check if they are testing the same thing. And we have to remove the unit tests eventually, anyway.


Sure! I'll remove those tests ASAP.

**Comment by NeerajGartia21** [comment]
> Note that https://github.com/prometheus/prometheus/pull/13999 might get merged very soon, and it will create merge conflicts. I think they will be easy to solve.

Yeah I was also thinking the same. But lets see how it goes after it gets merged.

**Comment by beorn7** [comment]
I still haven't found time for a detailed review. I will try later this week or early next week. My apologies.
In any case, I think we should wait for #13999, as it will be much easier to resolve merge conflicts in this PR than vice versa.

**Comment by NeerajGartia21** [comment]
> I still haven't found time for a detailed review. I will try later this week or early next week. My apologies.
In any case, I think we should wait for https://github.com/prometheus/prometheus/pull/13999, as it will be much easier to resolve merge conflicts in this PR than vice versa.

yeah, no problem. We should wait for #13999. Meanwhile I'll add the remaining tests.

**Comment by beorn7** [comment]
I plan to review this tomorrow (need one uninterrupted chunk of time for it…).

**Review by beorn7 — APPROVED** [review]
Nicely done. 

Also, more than 600 lines net removal!

**Comment by NeerajGartia21** [comment]
Thanks @beorn7 !

**Pull request merged**
